core/src/main/java/org/infinispan/interceptors/impl/InvalidationInterceptor.java

```diff
@@ -308,13 +315,14 @@ public Object visitPutMapCommand(InvocationContext ctx, PutMapCommand command) t
          // If the cache uses 2PC it's possible that the remotes will commit the invalidation and the originator rolls back,
          // but this does not impact consistency and the speed benefit is worth it.
          command = commandsFactory.buildPrepareCommand(txCtx.getGlobalTransaction(), Collections.singletonList(invalidateCommand), true);
+         command.setTopologyId(invalidateCommand.getTopologyId());
+      }
+      if (synchronous) {
```

Comment:
Yeah, it is annoying to have these extra checks in the interceptors. I still like the idea of keeping the sync and async cases separate in `RpcManager`, but maybe I should add something to `BaseRpcInterceptor` instead.